Modules/_json.c

```diff
@@ -628,40 +634,37 @@ py_encode_basestring(PyObject* Py_UNUSED(self), PyObject *pystr)
 }
 
 static void
-scanner_dealloc(PyObject *self)
+scanner_dealloc(PyScannerObject *self)
 {
+    PyTypeObject *tp = Py_TYPE(self);
     /* bpo-31095: UnTrack is needed before calling any callbacks */
     PyObject_GC_UnTrack(self);
     scanner_clear(self);
-    Py_TYPE(self)->tp_free(self);
+    freefunc free_func = PyType_GetSlot(tp, Py_tp_free);
```

Comment:
Even if the type is allocated on the heap, you can still use `Py_TYPE(self)->tp_free(self);`.

Same remark for similar changes below.